### PR TITLE
Document private items for binary crates by default

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
I suggested this change in default behavior [a long time ago](https://github.com/rust-lang/cargo/issues/1520#issuecomment-135215284) and [a year ago again](https://github.com/rust-lang/cargo/issues/1520#issuecomment-420616261). Both time, everyone seemed to agree that this is a good idea, so I thought I could just implement it.

This PR already changed the default behavior, but there are a few things we should talk about/I should do before merging:

- [x] ~~Do we *really* want this changed default behavior? If not, *why* not?~~ -> [yip](https://github.com/rust-lang/cargo/pull/7593#issuecomment-556482199)
- [x] Is changing this default behavior OK regarding backwards compatibility? -> [apparently](https://github.com/rust-lang/cargo/pull/7593#issuecomment-556482199)
- [x] ~~We should also probably add a `--document-only-public-items` flag or something like that if we change the default behavior. Otherwise users have no way to not document private items for binary crates. Right?~~ -> [We can do it later](https://github.com/rust-lang/cargo/pull/7593#issuecomment-557152039)
- [x] I should probably add some tests for this new behavior -> I did
- [ ] I don't like that I added `rustdoc_document_private_items` to `CompileOptions`, but this was the sanest way I could think of without rewriting a whole lot. Is this OK or how else would one do it? The flag would belong to `DocOptions`. But `compile` does not have access to that. Any ideas? Btw: If we also add `--document-only-private-items`, I would change the type from `bool` to `Option<bool>`.

### CODE REVIEW

**Comment by ehuss** [comment]
Thanks for the PR!  We discussed this briefly, and I don't think anyone had any objections to the general plan.  @Eh2406 will try to follow up with a review.

Just a side note, this should probably have a small test to validate the behavior.

**Comment by Eh2406** [comment]
Looks good to me with a test!
@ehuss what do you think of the `--document-only-private-items` flag?

**Comment by LukasKalbertodt** [comment]
Ooops, that was a typo! I of course meant `--document-only-public-items`. Otherwise the sentence "Otherwise users have no way to not document private items for binary crates." does not make any sense.

**Comment by ehuss** [comment]
I think it is fine without a flag.  If it seems like something people really want, we can add it later.

**Comment by LukasKalbertodt** [comment]
I added tests now. Are the two tests sufficient? Or not, what else to test?

This PR is now ready, I'd say :)

File: tests/testsuite/doc.rs [hunk 1]
```diff
@@ -1401,3 +1401,108 @@ fn doc_example() {
...
+        )
+        .run();
+
+    assert!(p.root().join("target/doc/foo/index.html").is_file());
+    assert!(p.root().join("target/doc/foo/fn.foo_pub.html").is_file());
+    assert!(p.root().join("target/doc/foo/fn.foo_priv.html").is_file());
+
+    assert!(p.root().join("target/doc/bar/index.html").is_file());
+    assert!(p.root().join("target/doc/bar/fn.bar_pub.html").is_file());
+    assert!(!p.root().join("target/doc/bar/fn.bar_priv.html").exists());
```
Comment (Eh2406) [inline tests/testsuite/doc.rs]: I think the tests are sufficient. But this behavior is not what I was expecting. For some reason I assumed that private details of dependencies would still be private. Was that just me? Is this the behavior we want?

Comment (LukasKalbertodt) [reply tests/testsuite/doc.rs]: Did you notice the `!` at the beginning, negating the assert? I.e. private items of dependencies are not documented.

Comment (Eh2406) [reply tests/testsuite/doc.rs]: oops. no sorry. :-(

Comment (LukasKalbertodt) [reply tests/testsuite/doc.rs]: No problem ^_^

I made this mistake countless times already, that's why I asked. `!` is just too small in a long assert.

**Comment by LukasKalbertodt** [comment]
Yeah :)

Not to be attention-seeking, but I think this deserves `relnotes`?

**Comment by davepacheco** [comment]
Thanks for doing this work!  To be precise, I think this change causes `cargo doc` to document private items for binary _targets_ in _all_ crates by default.  This is a nitty difference but I mention this because I spent a while trying to figure out whether I should expect to see private items in the library docs for a crate that also has two named binaries.  I'm not sure if you'd call this a "binary crate" or not, but anyway the result of this change is that you do get private items in the binary docs but not the library docs.  That makes sense.  I happened to want docs for private items in the libary, too, and I found a way to do that but ran into a small bug related to this issue: #8373.

**Comment by dtolnay** [comment]
@davepacheco I think the original description is right, though cargo and rustc extremely confusingly use different terms for the same concept.

> I think this change causes `cargo doc` to document private items for binary *targets* in all *crates* by default.

If you go by https://doc.rust-lang.org/1.44.0/cargo/reference/cargo-targets.html, this change documents private items for binary **targets** in all **packages** by default.

If you go by https://doc.rust-lang.org/1.44.0/book/ch07-01-packages-and-crates.html, this change documents private items for binary **crates** in all **packages** by default, with **targets** being a pretty different thing (https://doc.rust-lang.org/1.44.0/rustc/targets/index.html).

In either case, a collection of Rust crates/targets described by one Cargo.toml is called a Cargo package.

I think the use of "targets" for crates has been falling out of favor because targets already means the other thing, except maybe inside the Cargo codebase.

**Comment by davepacheco** [comment]
Thanks for that clarification!  Yeah, I got here from the [`cargo doc`](https://doc.rust-lang.org/cargo/commands/cargo-doc.html) documentation around target selection -- and you're right that it should be "package" where I said "crate".  The cargo book still seems to use "targets" pretty heavily (e.g., https://doc.rust-lang.org/cargo/reference/cargo-targets.html).

**Comment by Qix-** [comment]
Just as a note, I'm running into the case where I don't want private items documented for a binary. In fact, I think more refinement should be made to allow the types of visibility to be specified (e.g. crate public vs. module public) - but a discussion should happen there.

Initial PR at #10979.

**Pull request merged**